renormalizer/cv/finitet.py

```diff
@@ -429,86 +435,49 @@ def construct_X_qnmat(self, addlist, direction):
+            dag_qnbigl = inter_change(qnbigl)
+            dag_qnbigr = inter_change(qnbigr)
             if direction == "left":
-                list_qnbigl = qnbigl.ravel()
-                dag_qnbigl = []
-                for i in range(0, len(list_qnbigl), 2):
-                    dag_qnbigl.append([list_qnbigl[i + 1], list_qnbigl[i]])
-                dag_qnbigl = np.array(dag_qnbigl)
-                list_qnbigr = qnbigr.ravel()
-                dag_qnbigr = []
-                for i in range(0, len(list_qnbigr), 2):
-                    dag_qnbigr.append([list_qnbigr[i + 1], list_qnbigr[i]])
-                dag_qnbigr = np.array(dag_qnbigr).reshape(qnbigr.shape)
                 dag_qnbigr = np.moveaxis(dag_qnbigr, [0, 1], [1, 0])
             else:
-
-                list_qnbigr = qnbigr.ravel()
-                dag_qnbigr = []
-                for i in range(0, len(list_qnbigr), 2):
-                    dag_qnbigr.append([list_qnbigr[i + 1], list_qnbigr[i]])
-                dag_qnbigr = np.array(dag_qnbigr)
-
-                list_qnbigl = qnbigl.ravel()
-                dag_qnbigl = []
-                for i in range(0, len(list_qnbigl), 2):
-                    dag_qnbigl.append([list_qnbigl[i + 1], list_qnbigl[i]])
-                dag_qnbigl = np.array(dag_qnbigl).reshape(qnbigl.shape)
                 dag_qnbigl = np.moveaxis(dag_qnbigl, [1, 2], [2, 1])
         else:
-            dag_qnmat = np.moveaxis(dag_qnmat, [1, 2, 3, 4], [2, 1, 4, 3])
-            list_qnbigl = qnbigl.ravel()
-            dag_qnbigl = []
-            for i in range(0, len(list_qnbigl), 2):
-                dag_qnbigl.append([list_qnbigl[i + 1], list_qnbigl[i]])
-            dag_qnbigl = np.array(dag_qnbigl).reshape(qnbigl.shape)
-            dag_qnbigl = np.moveaxis(dag_qnbigl, [1, 2], [2, 1])
-            list_qnbigr = qnbigr.ravel()
-            dag_qnbigr = []
-            for i in range(0, len(list_qnbigr), 2):
-                dag_qnbigr.append([list_qnbigr[i + 1], list_qnbigr[i]])
-            dag_qnbigr = np.array(dag_qnbigr).reshape(qnbigr.shape)
-            dag_qnbigr = np.moveaxis(dag_qnbigr, [0, 1], [1, 0])
+            raise NotImplementedError
+            # we don't recommend 2-site CV-DMRG, which is a huge cost
 
         return dag_qnmat, dag_qnbigl, dag_qnbigr
 
     def condition(self, mat, qn):
-
-        list_qnmat = np.array(mat == qn).ravel()
-        mat_shape = list(mat.shape)
+        condition = (mat == qn)
+        mat_shape = list(condition.shape)
         del mat_shape[-1]
-        condition = []
-        for i in range(0, len(list_qnmat), 2):
-            if (list_qnmat[i] == 0) or (list_qnmat[i + 1] == 0):
-                condition.append(False)
-            else:
-                condition.append(True)
-        condition = np.array(condition)
+        condition = condition.all(axis=-1)
         condition = condition.reshape(mat_shape)
         return condition
```

Comment:
maybe you can give me some advice on improving this function, constitutes nearly 5% of the whole calculation apart from the rest 95% from cg. This function is used to compare the qn matrix with the required `qn`, `qn` is a list, eg. [0, 1] describing the quantum number of a local site CV mpo

Reply:
https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html

a = np.array([(1,1),(1,0)], dtype='i,i')
a == np.array((1,0),dtype='i,i')

Reply:
IMO it's not worthy of your time to pursue 5% possible speedup. If I were you I would rather play video games...